--- /tmp/bootchartd	2006-08-19 15:25:40.000000000 +0800
+++ bootchartd	2006-08-19 15:23:45.000000000 +0800
@@ -114,6 +114,10 @@
 # Wait for the boot process to end.
 wait_boot()
 {
+	TIMING_SCRIPT=/etc/init.d/bootcache-timing.sh
+	[ -x $TIMING_SCRIPT ] && source $TIMING_SCRIPT
+	[ "$UPTIME_LOG_TIMING" ] && { eval "$UPTIME_LOG_TIMING"; stop; return; }
+
 	local runlevel=$( sed -n 's/.*:\(.*\):initdefault:.*/\1/gp' /etc/inittab )
 
 	# The processes we have to wait for
